src/storage/AsyncStorage.js

```diff
@@ -0,0 +1,106 @@
+/* @flow strict-local */
+import { SQLDatabase } from './sqlite';
+
+/* eslint-disable no-underscore-dangle */
+
+// This is a Promise rather than directly a SQLDatabase because... well,
+// anything that wants to consume it needs to be prepared to wait in any
+// case, so will be calling something like `_db()` that returns a Promise.
+// And then that might as well return the same Promise every time, rather
+// than unwrapping it up front and wrapping it in a new Promise on each call.
+let dbSingleton: void | Promise<SQLDatabase> = undefined;
+
+// TODO: This needs a migration strategy!  How do we move the user's data
+//   from the old AsyncStorage to the new database?
+//
+//   On Android, we could bypass this by arranging to use the old thing's
+//   database and table name.  Might need some tweaking of expo-sqlite in
+//   order to control the database name fully enough.
+//
+//   But on iOS, it's not so simple.  Values over 1024 characters
+//   (RCTInlineValueThreshold, in RNCAsyncStorage.m) are written as separate
+//   files.  Values up to that threshold go into a single JSON blob (for the
+//   whole key-value store) that's written as one file.  So really our
+//   sanest path is probably to keep AsyncStorage as a dependency
+//   indefinitely, and use it to read the old data if the new doesn't exist.
```

Comment:
> Also, a nit: this file itself is named AsyncStorage, but I think the AsyncStorage you mean here is `@react-native-async-storage/async-storage`; do you think that needs to be disambiguated?

Looks like in a later commit I had the same thought, and made it "the legacy AsyncStorage". I'll backport that bit.

---

> Hm, yeah. Should we make a rough plan for how to approach AsyncStorage version upgrades in future? We'll lose stress-testing in our regular dev work for bugs triggered by upgrading it. Er, and I guess also for bugs triggered by _not_ upgrading it; for example, if our RN version advances such that something breaks in AsyncStorage.

Hmm. Yeah, I agree that this is a potential source of bugs in principle. I think it's not worth attempting to monitor for such a bug, though, for two reasons:
* The AsyncStorage library has been quite stable. As far as I've seen, there hasn't been any history of introducing bugs that would prevent reading old data.
* Our degree of dependence on it is going to rapidly diminish. A few weeks after we release with this change, most of our existing users will have upgraded, and then we'll never invoke the legacy AsyncStorage for them again. For new installs, the only thing we'll need from `legacyAsyncStorage` is for it to successfully return an empty list from `getAllKeys()`, which hopefully keeps it pretty unlikely that it would break.

  Then even for the remaining existing users belatedly upgrading over time, our needs are only slightly more than that: in particular purely read-only, never writing to the legacy storage again. Reads are generally a lot lower-risk than writes, so that should also keep the chances low of a bug that would affect us.

In the long term, my plan is:
```
  async _migrateFromLegacyAsyncStorage(tx) {
    // TODO: It would be nice to reduce the LegacyAsyncStorage dependency to
    //   be read-only -- in particular, for new installs to stop creating an
    //   empty legacy store, which on Android happens just from initializing
    //   the legacy AsyncStorage module.  This will basically mean vendoring
    //   the library into src/third-party/, and then stripping out
    //   everything not needed for read-only use.
```
but I think we can do that sometime in the future when we feel like reducing dependencies.